zkevm-circuits/src/pi_circuit.rs

```diff
@@ -256,113 +393,276 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-        // offset
-        meta.create_gate("block_table[i] = raw_public_inputs[offset + i]", |meta| {
-            let q_block_table = meta.query_selector(q_block_table);
-            let block_value = meta.query_advice(block_table.value, Rotation::cur());
-            let rpi_block_value = meta.query_advice(raw_public_inputs, Rotation::cur());
-            vec![q_block_table * (block_value - rpi_block_value)]
+                cb.gate(
+                    meta.query_selector(q_rpi_byte_not_end) *
+                    not::expr(meta.query_selector(q_rpi_byte_not_end))
+                )
+            },
+        );
+
+        // 3: q_rpi_value_start is bool && q_digest_value_start is bool && is_value_rlc_cur is
+        // bool
+        meta.create_gate("fixed column is bool", |meta| {
+            let q_rpi_value_start_cur = meta.query_fixed(q_rpi_value_start, Rotation::cur());
+            let q_digest_value_start_cur = meta.query_fixed(q_digest_value_start, Rotation::cur());
+            let is_value_rlc_cur = meta.query_fixed(is_value_rlc, Rotation::cur());
+
+            vec![or::expr([
+                q_rpi_value_start_cur.clone() * (1.expr() - q_rpi_value_start_cur),
+                q_digest_value_start_cur.clone() * (1.expr() - q_digest_value_start_cur),
+                is_value_rlc_cur.clone() * (1.expr() - is_value_rlc_cur),
+            ])]
         });
 
-        let offset = BLOCK_LEN + 1 + EXTRA_LEN;
-        let tx_table_len = max_txs * TX_LEN + 1;
+        // 4: rpi_value_rlc[i+1] = rpi_value_rlc[i] * (is_value_rlc[i] ? evm_rand : byte_pow_base )
+        // + rpi_bytes[i+1]
+        meta.create_gate(
+            "rpi_value_rlc[i+1] = rpi_value_rlc[i] * (is_value_rlc[i] ? evm_rand : byte_pow_base) + rpi_bytes[i+1]",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::default();
+                let q_rpi_byte_not_end_cur = meta.query_selector(q_rpi_byte_not_end);
+                let q_rpi_value_start_next = meta.query_fixed(q_rpi_value_start, Rotation::next());
+                let is_value_rlc_cur = meta.query_fixed(is_value_rlc, Rotation::next());
+                let rpi_value_rlc_next = meta.query_advice(rpi_value_rlc, Rotation::next());
+                let rpi_value_rlc_cur = meta.query_advice(rpi_value_rlc, Rotation::cur());
+                let rpi_bytes_next = meta.query_advice(rpi_bytes, Rotation::next());
+                let r = select::expr(is_value_rlc_cur, challenges.evm_word(), BYTE_POW_BASE.expr());
+
+                cb.require_equal(
+                    "rpi_value_rlc[i+1] = rpi_value_rlc[i]* r + rpi_bytes[i+1]",
+                    rpi_value_rlc_next,
+                    rpi_value_rlc_cur * r + rpi_bytes_next,
+                );
 
-        //  0.3 Tx table -> {tx_id, index, value} column match with raw_public_inputs
-        // at expected offset
+                cb.gate(not::expr(q_rpi_value_start_next) * q_rpi_byte_not_end_cur)
```

Comment:
Suggestion: `q_rpi_value_start` is 0 or 1.
When it's 1 we start the RLC/LC accumulation, when it's 0 we continue the RLC/LC accumulation.
I see this as 2 branches (one is gate 4, the other is gate 5).  I think it would be easier to reason about if the gates selector expressions are `q_rpi_byte_not_end_cur` in gate 5 (like it currently is) and `not(q_rpi_byte_not_end_cur)` in gate 4.

So I would rewrite gate 4 like this:
```python
if not(q_rpi_value_start_cur):
  assert rpi_value_rlc[i] == rpi_value_rlc[i-1] * r + rpi_bytesp[i]
```

To handle the last value case I would just add a `q_enable` which is 1 in all rows that contain data, and multiply that selector to all gates.

This way to write gates makes it easier to follow all possible cases and I think it makes it simpler to reason that there are no underconstraint cases.

Reply:
This suggestion really nice, fixed all of it